crypto3/libs/multiprecision/include/nil/crypto3/multiprecision/ops/import_export.hpp

```diff
@@ -0,0 +1,210 @@
+         * fits into one limb (machine word).
+         */
+        template<std::size_t Bits, typename Unsigned>
+        void assign_bits(big_uint<Bits>& val, Unsigned bits, std::size_t bit_location,
+                         std::size_t chunk_bits,
+                         const std::integral_constant<bool, true>& /*unused*/) {
+            using limb_type = typename big_uint<Bits>::limb_type;
+            //
+            // Check for possible overflow.
+            //
+            NIL_CO3_MP_ASSERT(!((bit_location >= limb_bits) && bits));
+
+            limb_type mask = chunk_bits >= limb_bits
+                                 ? ~static_cast<limb_type>(0u)
+                                 : (static_cast<limb_type>(1u) << chunk_bits) - 1;
+            limb_type value = (static_cast<limb_type>(bits) & mask) << bit_location;
+            *val.limbs() |= value;
+
+            //
+            // Check for overflow bits:
+            //
+            bit_location = limb_bits - bit_location;
+
+            NIL_CO3_MP_ASSERT(
+                !((bit_location < sizeof(bits) * CHAR_BIT) && (bits >>= bit_location)));
+        }
+
+        template<std::size_t Bits>
+        std::uintmax_t extract_bits(const big_uint<Bits>& val, std::size_t location,
+                                    std::size_t count) {
+            std::size_t limb = location / limb_bits;
+            std::size_t shift = location % limb_bits;
+            std::uintmax_t result = 0;
+            std::uintmax_t mask = count == std::numeric_limits<std::uintmax_t>::digits
+                                      ? ~static_cast<std::uintmax_t>(0)
+                                      : (static_cast<std::uintmax_t>(1u) << count) - 1;
+            if (count > (limb_bits - shift)) {
+                result = extract_bits(val, location + limb_bits - shift,
+                                      count - limb_bits + shift);
+                result <<= limb_bits - shift;
+            }
+            if (limb < val.limbs_count()) {
+                result |= (val.limbs()[limb] >> shift) & mask;
+            }
+            return result;
+        }
+
+        template<std::size_t Bits, typename Iterator>
+        big_uint<Bits>& import_bits_generic(big_uint<Bits>& result, Iterator i, Iterator j,
+                                            std::size_t chunk_size = 0, bool msv_first = true) {
+            big_uint<Bits> newval;
+
+            using value_type = typename std::iterator_traits<Iterator>::value_type;
+            using difference_type = typename std::iterator_traits<Iterator>::difference_type;
+            using size_type = typename std::make_unsigned<difference_type>::type;
+
+            if (!chunk_size) {
+                chunk_size = std::numeric_limits<value_type>::digits;
+            }
+
+            size_type limbs = std::distance(i, j);
+            size_type bits = limbs * chunk_size;
+
+            // TODO(ioxid): this breaks marshalling tests
```

Comment:
I think we need another check here, if I understand correctly. Very often we have 255 bit numbers, or 381 bit numbers. In this case marshalling may give you 4 limbs of 64 bits and 6 limbs of 64 bits respectively, since 255 and 381 are not divisible by 64. What you should check here is that there extra bits are ALL ZERO! 

And better have a throw here, not an assert, since it may happen in production, if someone intentionally gives us a wrong marshalled file as input.

Reply:
Added this to the comment. I think I can do this after initial merge.